satellite/marketing/marketingweb/server.go

```diff
@@ -6,43 +6,79 @@ package marketingweb
 import (
 	"context"
 	"html/template"
+	"log"
 	"net"
 	"net/http"
 	"path/filepath"
+	"reflect"
+	"time"
 
 	"github.com/gorilla/mux"
+	"github.com/gorilla/schema"
 	"github.com/zeebo/errs"
 	"go.uber.org/zap"
 	"golang.org/x/sync/errgroup"
+
+	"storj.io/storj/satellite/marketing"
 )
 
-// Error is satellite marketing error type
-var Error = errs.Class("satellite marketing error")
+var (
+	// Error is satellite marketing error type
+	Error   = errs.Class("satellite marketing error")
+	decoder = schema.NewDecoder()
+)
 
-// Config contains configuration for marketing offersweb server
+// Config contains configuration for marketingweb server
 type Config struct {
 	Address   string `help:"server address of the marketing Admin GUI" default:"127.0.0.1:8090"`
 	StaticDir string `help:"path to static resources" default:""`
 }
 
 // Server represents marketing offersweb server
 type Server struct {
-	log *zap.Logger
-
-	config Config
-
-	listener net.Listener
-	server   http.Server
-
+	log         *zap.Logger
+	config      Config
+	listener    net.Listener
+	server      http.Server
+	service     *marketing.Service
 	templateDir string
 	templates   struct {
 		home          *template.Template
 		pageNotFound  *template.Template
 		internalError *template.Template
+		badRequest    *template.Template
 	}
 }
 
-// commonPages returns templates that are required for everything.
+// offerSet provides a separation of marketing offers by type.
+type offerSet struct {
+	RefOffers, FreeCredits []marketing.Offer
+}
+
+// init safely registers convertStringToTime for the decoder.
+func init() {
+	decoder.RegisterConverter(time.Time{}, convertStringToTime)
+}
+
+// organizeOffers organizes offers by type.
+func organizeOffers(offers []marketing.Offer) offerSet {
```

Comment:
do you have a unit test for this ?

Reply:
This is a helper function for `GetOffers` which is unit-tested.